src/biz/workspace/publish_dup.rs

```diff
@@ -0,0 +1,959 @@
+      duplicated_db_main_view: HashMap::new(),
+      duplicated_db_view: HashMap::new(),
+
+      pg_pool,
+      collab_storage,
+      duplicator_uid: dest_uid,
+      dest_workspace_id,
+      dest_view_id,
+    }
+  }
+
+  async fn duplicate(mut self, publish_view_id: &str) -> Result<(), AppError> {
+    // new view after deep copy
+    // this is the root of the document/database duplicated
+    let mut root_view = match self.deep_copy(gen_view_id(), publish_view_id).await? {
+      Some(v) => v,
+      None => {
+        return Err(AppError::RecordNotFound(
+          "view not found, it might be unpublished".to_string(),
+        ))
+      },
+    };
+    root_view.parent_view_id.clone_from(&self.dest_view_id);
+
+    // destructuring self to own inner values, avoids cloning
+    let PublishCollabDuplicator {
+      collab_storage,
+      duplicated_refs: _,
+      duplicated_db_main_view: _,
+      duplicated_db_view: _,
+      mut views_to_add,
+      workspace_databases,
+      collabs_to_insert,
+      ts_now: _,
+      pg_pool,
+      duplicator_uid,
+      dest_workspace_id,
+      dest_view_id: _,
+    } = self;
+
+    // insert all collab object accumulated
+    // for self.collabs_to_insert
+    let mut txn = pg_pool.begin().await?;
+    for (oid, (collab_type, encoded_collab)) in collabs_to_insert.into_iter() {
+      collab_storage
+        .insert_new_collab_with_transaction(
+          &dest_workspace_id,
+          &duplicator_uid,
+          CollabParams {
+            object_id: oid.clone(),
+            encoded_collab_v1: encoded_collab,
+            collab_type,
+            embeddings: None,
+          },
+          &mut txn,
+        )
+        .await?;
+    }
+
+    // update database if any
+    if !workspace_databases.is_empty() {
+      let ws_db_oid = select_workspace_database_oid(&pg_pool, &dest_workspace_id.parse()?).await?;
+      let mut ws_db_collab = {
+        let ws_database_ec = get_latest_collab_encoded(
+          collab_storage.clone(),
+          &duplicator_uid,
+          &dest_workspace_id,
+          &ws_db_oid,
+          CollabType::WorkspaceDatabase,
+        )
+        .await?;
+        collab_from_doc_state(ws_database_ec.doc_state.to_vec(), &ws_db_oid)?
+      };
+
+      let ws_db_body = WorkspaceDatabaseBody::new(&mut ws_db_collab);
+
+      let (ws_db_updates, updated_ws_w_db_collab) = tokio::task::spawn_blocking(move || {
+        let ws_db_updates = {
+          let mut txn_wrapper = ws_db_collab.transact_mut();
+          for (db_collab_id, linked_views) in &workspace_databases {
+            ws_db_body.add_database(&mut txn_wrapper, db_collab_id, linked_views.clone());
+          }
+          txn_wrapper.encode_update_v1()
+        };
+        let updated_ws_w_db_collab = collab_to_bin(&ws_db_collab, CollabType::WorkspaceDatabase);
+        (ws_db_updates, updated_ws_w_db_collab)
+      })
+      .await?;
+
+      collab_storage
+        .insert_new_collab_with_transaction(
```

Comment:
how about using `insert_new_collab`? it will save the collab in memory and the write to disk in the background.